### ISSUE TITLE
[FR] new --and, --not, and --bool operators

### ISSUE THREAD

**Issue by genivia-inc** [Genivia/ugrep#69]
The `--not` operator is available since version 2.5 to exclude matching lines that match the specified `--not` regex patterns, e.g. `-e PATTERN --not -e PATTERN1 -e PATTERN2` matches (multi-)lines with `PATTERN` but not when `PATTERN1` and `PATTERN2` also match.

A new `--and` operator would be nice to have to specify two or more patterns that must match a matching line.

When implemented, this feature can be the basis of a special search-engine like pattern syntax for NOT and AND operators, enabled with a new command line option, see the discussion #31. With this option, a regex pattern `P Q` matches both `P` and `Q` and `P -Q` matches `P` if the line has no `Q`.

The proper design for `--and` requires some retooling. We need to check the performance impact of this, which is should be minimal.

We are also looking into a new `--bool` option to specify Boolean search queries with `AND` (or just space), `OR` (or `|`), `NOT` (or `-`), `" "` quotes (exact match), and `(` `)` for grouping. This means that the Boolean search query patterns are similar to Google queries.

**Comment by genivia-inc** [comment]
One of the upcoming ugrep releases adds new `--and`, `--if`, `--not` connectives to refine the search with constraints.

The current proposal for the design and implementation is as follows (subject to change).

The `--and` connective combines two or more patterns that must all match, with possible alternatives:
~~~
ug -e A -e B --and -e C -e D --and -e E -e F
~~~
This searches for (multi-)lines matching (A OR B) AND (C OR D) AND (E or F). For example, a line containing B, C, E, and F is matched. This CNF form (conjunctive normal form) is similar to git-grep's approach:
~~~
git-grep -e A -e B --and \( -e C -e D \) --and \( -e E -e F \)
~~~
but without the parenthesis.

The search results with `--and` shows lines with all matches color-highlighted. So for example, a line with B, C and F shows B, C, E and F highlighted.

The `--if` connective is similar to `--and`, but does not highlight all matches, only the first specified before the `--if`. This connective adds a condition on the search without affecting the output of the search, whereas `--and` affects the search results by matching and highlighting all patterns specified:
~~~
ug -e A -e B --if -e C -e D --if -e E -e F
~~~
This searches for (multi-)lines matching (A OR B) AND (C OR D) AND (E or F). For example, a line containing B, C and F is matched. But only B is considered the pattern to match and highlight, given that the specified constraints are met by C and F.

The `--not` connective excludes the specified patterns from matches:
~~~
ug -e A -e B --not -e C -e D --not -e E -e F
~~~
This searches for (multi-)lines matching (A OR B) AND NOT (C OR D) AND NOT (E OR F). For example, a line containing B and C but not D, E, F is matched (i.e. both AND NOT clauses are applicable; if any one fails then there is a match).

This upgraded `--not` connective can be used with `-P` (Perl matching), which is currently not possible in v3.0. The current v3.0 implementation also does not obey multiple `--not` to obtain the AND NOT semantics described above.

The `--and` operator can be used together with `--not`, but this combination is superfluous:
~~~
ug  -e A -e B --and --not -e C -e D --and --not -e E -e F 
ug  -e A -e B --not -e C -e D --not -e E -e F 
~~~
are two forms for the same query (A OR B) AND NOT (C OR D) AND NOT (E or F)

The `--and` operator can also be used after one or more `--not` connectives. For example:
~~~
ug --not -e C -e D --not -e E -e F --and -e A -e B 
ug  -e A -e B --not -e C -e D --not -e E -e F 
~~~
are two forms for the same query NOT (C OR D) AND NOT (E or F) AND (A OR B) == (A OR B) AND NOT (C OR D) AND NOT (E or F).

Note that we do not need an `--or` connective like git-grep has, which would be potentially confusing to use. In fact, a sequence of two or more `-e PATTERN` forms a disjunction. An `--or` operator is not needed.

Because the connectives can be used to formulate a query in CNF, any query with nested AND, OR, NOT connectives can be transformed into CNF. However, the ugrep release plans do not yet offer an automatic translation to CNF, which may be considered in a later project phase.

In general, all propositional connectives are used with `-e PATTERN`, but also`-N PATTERN` can be used. The `-N PATTERN` patterns are negatives, i.e. a negative pattern is essentially invisible to the search. This differs from the `--not -e PATTERN` that excludes a line matching `PATTERN' anywhere on the line.

If you have suggestions or concerns, please reply to this post!

**Comment by genivia-inc** [comment]
We're making good progress with the Boolean operators. This is taking a bit longer because of the alternative design choices to evaluate for the command line options. We will also add a new pattern syntax with a new option `--bool` (tentative name). This option allows you to specify Boolean patterns with the connectives AND, OR, and NOT. The `--bool` pattern syntax has some similarities with the Goggle search engine query syntax and will support the following operators:

pattern | alternative | result
------- | ----------- | -------
`x y`   | `x AND y`   | matches lines with both x and y
`x\|y`  | `x OR y`    | matches lines with x or y
`-x`    | `NOT x`     | matches lines that do not contain x
`( )`   |             | Boolean expression grouping
`“x”`   |             | match x literally, do not interpret `AND`, `OR`, `NOT`, `\|`, `-`, and `()` in string x

Terms `x` and `y` are any strings without spacing that do not start with the special symbols `|`, `-`, and `(` (use a `\` escape to match these);

`|` and `OR` are the same and take precedence over `AND`: `x y|z` == `x (y|z)`;

`-` and `NOT` are the same and take precedence over `OR`: `-x|y` == `(-x)|y`.

Some simple examples of `--bool PATTERN`:

`hot dog` matches lines matching both `hot` and `dog`;
`hotdog|taco place` matches lines matching `place` and then also `hotdog` or `taco`;
`hotdog|taco place -diner` same as above, but exclude lines matching `diner`;
`diner|(fast food -bad)` matches lines matching `diner` or lines matching both `fast` and `food` but not `bad`.
`"fast food"|diner -bad -old` matches lines matching the string `fast food` or lines matching `diner`, excluding lines with `bad` or `old`;
`"fast food"|diner -(bad|old)` same as above (because their conjunctive normal forms are the same).

The update will be released soon when all the machinery is in place and thoroughly tested.

**Comment by genivia-inc** [comment]
Good news. The implementation is finalizing. We're progressing into the testing phase. We plan to release 3.1 sometime soon, depending on our rigorous testing results.

The new options are `--bool SEARCH` to specify a Google-like `SEARCH` query with the operators described in the table posted earlier, and `--and` and `--not` (and a `--andnot` shorthand). These connectives allow you to specify additional constraints as command line arguments.

The `--and` connective applies to one or more `-e` patterns that are OR-ed. For example, `-e 'A|B' --and -e 'X|Y'` is the same as `-e A -e B --and -e X -e Y`.

The `--not` connective inverts a pattern match. For example, `-e A --and -e B --not -e C` matches lines with `A` and `B` without `C`. Note that `-e A --and --not -e B` matches lines with `A` without `B`. This can be shortened to `-e A --andnot -e B`. Note that `-e A --not -e B` is not the same! Because this matches lines with `A` or lines without `B`, so there may not be any `A` matched in the results. This behavior is the same as Git grep.

The `-e` after a `--and`, `--andnot`, and `--not` is optional.

Looking forward to testing and updating the manual and man page.

**Comment by genivia-inc** [comment]
Implemented in v3.1.
